main.go

```diff
@@ -169,36 +373,115 @@ func NewInternodeDialContext(dialTimeout time.Duration) DialContext {
 
 // Reader to read random data.
 type netperfReader struct {
-	doneCh <-chan struct{}
-	buf    []byte
+	buf []byte
+
+	Host   string
+	Addr   string
+	count  atomic.Uint64
+	TX     atomic.Uint64
+	TTFBH  atomic.Int64
+	TTFBL  atomic.Int64
+	Errors atomic.Uint64
+}
+
+type AsyncReader struct {
+	pr             *netperfReader
+	i              int64 // current reading index
+	prevRune       int   // index of previous rune; or < 0
+	TTFBRegistered bool
+	Start          time.Time
+	ctx            context.Context
 }
 
-func (m *netperfReader) Read(b []byte) (int, error) {
-	select {
-	case <-m.doneCh:
-		return 0, io.EOF
-	default:
+func (a *AsyncReader) Read(b []byte) (n int, err error) {
+	if !a.TTFBRegistered {
+		a.TTFBRegistered = true
+		since := time.Since(a.Start).Milliseconds()
+		if since > a.pr.TTFBH.Load() {
+			a.pr.TTFBH.Store(since)
+		}
```

Comment:
You are "racing". Here is a safe update:
```
		x := a.pr.TTFBH.Load()
		for since > x && !a.pr.TTFBH.CompareAndSwap(x, since) {
			x = a.pr.TTFBH.Load()
		}
```